### PR TITLE
Annotate `Type.GetInterfaceMap` to guarantee trim compatibility

### LINKED ISSUE #46912: Type.GetInterfaceMap is not annotated for trimming
I think illink will already make sure we don't trim the interface type from the interface list, but illink might still consider the inteface methods to be a fair game. To prevent behavioral differences I think we need to annotate the `ifaceType` parameter as requiring all public methods.

This was flagged by the analysis in NativeAOT since the NativeAOT implementation of this API is done with regular C#.

### PR DESCRIPTION
**Pull request by vitek-karas**
Fixes https://github.com/dotnet/runtime/issues/46912

### CODE REVIEW

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs [hunk 1]
```diff
@@ -2667,7 +2667,7 @@ public override MemberInfo[] GetMembers(BindingFlags bindingAttr)
             return members;
         }
 
-        public override InterfaceMapping GetInterfaceMap(Type ifaceType)
+        public override InterfaceMapping GetInterfaceMap([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods)] Type ifaceType)
```
Comment (eerhardt) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:2670]: Why `NonPublicMethods`?

**Review by eerhardt — APPROVED** [review]
I just had one question, which I'm sure you have a valid answer for 😄 . Once that is answered, LGTM.

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs (hunk 1 above)
Comment (vitek-karas) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:2670]: As per Michal's example:
```C#
using System;

interface IFoo
{
    public void DoWork() => Console.WriteLine("UU");
    internal void OtherWork() => Console.WriteLine("OO");
}

class Base : IFoo
{
    static void Main()
    {
        foreach (var v in typeof(Base).GetInterfaceMap(typeof(IFoo)).TargetMethods)
        {
            Console.WriteLine(v);
        }
    }
}
```

This actually prints out both methods, the public `DoWork` as well as the internal `OtherWork`... so to keep this working the same after trimming - we need to preserve both public and non-public methods.

Comment (eerhardt) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:2670]: Thanks for the explanation.

**Pull request merged**